api/analyzedphenotypes.columnheaders.api.inc

```diff
@@ -0,0 +1,277 @@
+      );
+    }
+  }
+
+  return $columnheaders;
+}
+
+
+///////
+
+
+/**
+ * CONTROL/SWITCHBOARD.
+ * Pass a request to a corresponding function.
+ * @TODO: leave out when api reorganize/update is complete and use functions below.
+ */
+function analyzedphenotypes_columnsprop($property = null, $parameter = null) {
+  if (is_null($property)) {
+    // Default.
+    $property = 'default';
+  }
+
+  // Process request.
+  $command = array(
+    'default' => array(
+      'func'  => 'ap_create_columnheaders',
+      'param' => 'columns_datatype',
+    ),
+    'combination' => array(
+      'func'  => 'ap_create_columnheaders',
+      'param' => 'columns_combination',
+    ),
+    'downloadable' => array(
+      'func'  => 'ap_create_columnheaders',
+      'param' => 'columns_downloadable',
+    ),
+    'column_to_field' => array(
+      'func'  => 'ap_create_columnheaders',
+      'param' => 'columns_downloadable_tofield',
+    ),
+    'match_combination' => array(
+      'func'  => 'ap_match_columncombination',
+      'param' => $parameter,
+    ),
+    'rfriendly' => array(
+      'func'  => 'ap_convert_rcolumnheader',
+      'param' => $parameter['column_header'],
+    ),
+  );
+
+  call_user_func($command[$property]['func'], $command[$property]['param']);
+}
+
+
+///////
+
+
+/**
+ * FUNCTION:
+ * Test combnination of column headers against matching column headers from a data point/row to see if it
+ * exsits in chado.phenotype and/or chado.phenotypeprop tables.
+ * Dependencies: ap_create_columnheaders().
+ *
+ * @param $datarow
+ *   An array, a single row of data from a tsv data sheet. Data/cell filtered to only include
+ *   cells matching column headers in column combnination set.
+ *
+ * @return
+ *   Record id number when match found and null when no match found.
+ */
+function ap_match_columncombination($datarow) {
+  // Fetch combination of columns to test uniqueness of record.
+  $combination_columns = ap_create_columnheaders('combination_columns');
+
+  $sysvar = ap_get_variablenames(
+    array('variablename' => 'varset'),
+    array('set' => 'terms')
+  );
+
+  foreach($sysvar as $i => $var) {
+    $sysvar_terms[$i] = variable_get($var);
+  }
+
+  $result = chado_query(
+    "SELECT phenotype.phenotype_id
+     FROM
+       {phenotype} AS phenotype
+       LEFT JOIN {cvterm}        AS trait_name ON phenotype.attr_id = trait_name.cvterm_id
+       LEFT JOIN {stock}         AS germplasm_accession USING (stock_id)
+       LEFT JOIN {phenotypeprop} AS year USING (phenotype_id)
+       LEFT JOIN {phenotypeprop} AS location USING (phenotype_id)
+       LEFT JOIN {phenotypeprop} AS replicate USING (phenotype_id)
+     WHERE
+       year.type_id = :year_cvterm
+       AND location.type_id  = :location_cvterm
+       AND replicate.type_id = :replicate_cvterm
+       AND trait_name.name || germplasm_accession.uniquename || year.value || location.value || replicate.value = :combination"
```

Comment:
My are you concatenating here for the comparison? I think it would be faster postgresql-wise to compare each individually and also better handle edge cases such as numbers at the end of the location that mimic a replicate, etc.
```sql
  AND trait_name.name = :trait
  AND germplasm_accession.uniquename = :germ
  AND year.value = :year
  AND location.value = :location
  AND replicate.value = :rep

Reply:
I have revised this api. We have this idea of unique combination and so I coded based on this idea - combination/combine, hence I concatenated :)